### DISCUSSION TITLE
Is xformers disabled in latest commit?

### DISCUSSION THREAD

**Discussion by linkksng** [vladmandic/sdnext#64] [Q&A] [answered]
Since PyTorch 2.0 is GA, I have been using it with xformers (using command line arguments `--xformers`), and getting similar performance with Pytorch 1.13. SDP `--opt-sdp-attention` gives much worse performance. (2.1 it/s vs 0.38 it/s, but that would be a separate issue on its own...)
After updating to latest commit a13aeccd, xformers seems to be disabled. Performance is now down to 0.38 it/s which is the same as using SDP. When I run system info benchmark, the result shows optimization being "sdp" instead of "xformers".
Is the latest commit automatically disabling xformers in favour of SDP?

OS: Linux Mint 21.1 x64
GPU: GTX 1070 (8 GB VRAM)
CPU: i7-7700K

**Comment by vladmandic** [comment, marked as answer]
yes, latest version prefers `sdp` over `xformers` by default.  
if sdp is not available it will try xformers and xformers are not available then doggetx.  

i've just added extra cmd flag to disable sdp if you don't want to use it: `--disable-sdp-attention`

**Reply by linkksng** [reply]
Thanks for the quick response and update!
Confirm that `--disable-sdp-attention --xformers` re-enables xformers and the corresponding performance

**Reply by Centurion-Rome** [reply]
also worked for me, thnx!

**Reply by Woisek** [reply]
I can't confirm this. I added --disable-sdp-attention --xformers but xformers show N/A in the footer zone of the GUI.

**Comment by Woisek** [comment]
I also may add, when starting up, it writes:
Launching Web UI with arguments: --xformers --api --administrator --opt-split-attention --theme=dark
No module 'xformers'. Proceeding without it.

Altough xformes should be installed. At least it says so when I do pip install xformers ...
Any suggestions?

**Comment by vladmandic** [comment]
fyi, major update posted, see #99 for details. as a result of that, all cross-optimization flags are gone and now you should configure it from the ui itself.

**Reply by Woisek** [reply]
Cool stuff! What about flags like --auth and --administrator, needed for PS plugins?

**Reply by vladmandic** [reply]
full list of remaining is listed. not sure which auth is needed by ps plugin as auth is set on server, extensions just use it. same for admin - why would extension need that?!

**Reply by Woisek** [reply]
OK, update for the ... update 🤪:
Doesnt start, complains about xformers.

Already up to date.
venv "I:\Super SD 2.0\automatic\venv\Scripts\Python.exe"
19:33:30-284557 INFO     Python 3.10.6 on Windows
19:33:30-289543 INFO     Running setup
19:33:30-293533 INFO     Installing requirements
19:33:30-533916 INFO     Version: 910f047c Wed Apr 12 12:51:11 2023 -0400
19:33:30-906070 WARNING  Latest available version: 2023-04-12T16:51:11Z
19:33:30-909038 INFO     Updating Wiki
19:33:31-535525 ERROR    Error running git with args: -C "I:\Super SD 2.0\automatic\wiki\origin-wiki" branch
19:33:31-538492 WARNING  Unknown branch for: I:\Super SD 2.0\automatic\wiki\origin-wiki
19:33:31-571435 ERROR    Error running git with args: -C "I:\Super SD 2.0\automatic\wiki\origin-wiki" pull --rebase
                         --autostash
19:33:31-604350 ERROR    Error running git with args: -C "I:\Super SD 2.0\automatic\wiki\origin-wiki" branch
19:33:31-606311 INFO     Installing packages
┌───────────────────────────────────────── Traceback (most recent call last) ─────────────────────────────────────────┐
│ I:\Super SD 2.0\automatic\launch.py:87 in <module>                                                                  │
│                                                                                                                     │
│   86 if __name__ == "__main__":                                                                                     │
│ > 87     setup.run_setup(False)                                                                                     │
│   88     setup.set_environment()                                                                                    │
│                                                                                                                     │
│ I:\Super SD 2.0\automatic\setup.py:336 in run_setup                                                                 │
│                                                                                                                     │
│   335     check_version()                                                                                           │
│ > 336     install_packages()                                                                                        │
│   337     install_repositories()                                                                                    │
│                                                                                                                     │
│ I:\Super SD 2.0\automatic\setup.py:174 in install_packages                                                          │
│                                                                                                                     │
│   173     install(openclip_package)                                                                                 │
│ > 174     install(f'--no-deps {xformers_package}')                                                                  │
│   175                                                                                                               │
│                                                                                                                     │
│ I:\Super SD 2.0\automatic\setup.py:86 in install                                                                    │
│                                                                                                                     │
│    85                                                                                                               │
│ >  86     if not installed(package):                                                                                │
│    87         pip(f"install --upgrade {package}")                                                                   │
│                                                                                                                     │
│ I:\Super SD 2.0\automatic\setup.py:59 in installed                                                                  │
│                                                                                                                     │
│    58             ok = ok and spec is not None                                                                      │
│ >  59             version = pkg_resources.get_distribution(p[0]).version                                            │
│    60             if ok and len(p) > 1:                                                                             │
│                                                                                                                     │
│ I:\Super SD 2.0\automatic\venv\lib\site-packages\pkg_resources\__init__.py:478 in get_distribution                  │
│                                                                                                                     │
│    477     if isinstance(dist, Requirement):                                                                        │
│ >  478         dist = get_provider(dist)                                                                            │
│    479     if not isinstance(dist, Distribution):                                                                   │
│                                                                                                                     │
│ I:\Super SD 2.0\automatic\venv\lib\site-packages\pkg_resources\__init__.py:354 in get_provider                      │
│                                                                                                                     │
│    353     if isinstance(moduleOrReq, Requirement):                                                                 │
│ >  354         return working_set.find(moduleOrReq) or require(str(moduleOrReq))[0]                                 │
│    355     try:                                                                                                     │
│                                                                                                                     │
│ I:\Super SD 2.0\automatic\venv\lib\site-packages\pkg_resources\__init__.py:909 in require                           │
│                                                                                                                     │
│    908         """                                                                                                  │
│ >  909         needed = self.resolve(parse_requirements(requirements))                                              │
│    910                                                                                                              │
│                                                                                                                     │
│ I:\Super SD 2.0\automatic\venv\lib\site-packages\pkg_resources\__init__.py:795 in resolve                           │
│                                                                                                                     │
│    794                         requirers = required_by.get(req, None)                                               │
│ >  795                         raise DistributionNotFound(req, requirers)                                           │
│    796                 to_activate.append(dist)                                                                     │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
DistributionNotFound: The 'xformers' distribution was not found and is required by the application
Drücken Sie eine beliebige Taste . . .

Suggestions? 🙂

**Reply by Woisek** [reply]
> full list of remaining is listed. not sure which auth is needed by ps plugin as auth is set on server, extensions just use it. same for admin - why would extension need that?!

Not extensions, a/the plugins for Photoshop. Like this one I use -> https://github.com/isekaidev/stable.art

**Reply by vladmandic** [reply]
> > full list of remaining is listed. not sure which auth is needed by ps plugin as auth is set on server, extensions just use it. same for admin - why would extension need that?!
> 
> Not extensions, a/the plugins for Photoshop. Like this one I use -> https://github.com/isekaidev/stable.art

even worse. plugin is a pure consumer of api. it could be internal api or actual public api, but in both case, it has nothing to do with auth set on sd server - it should use it as-is. and admin priviledges? why would ps plugin require that sd server runs with admin priviledges? thats just ridiculous.

**Reply by vladmandic** [reply]
regarding xformers, see separate thread.

**Reply by Woisek** [reply]
> even worse. plugin is a pure consumer of api. it could be internal api or actual public api, but in both case, it has nothing to do with auth set on sd server - it should use it as-is. and admin priviledges? why would ps plugin require that sd server runs with admin priviledges? thats just ridiculous.

Sorry, I'm just a User who uses this plugin. 🙄
I absolutely understand it from your point of view, and I guess this is something coming from Adobe so that PS and WebUI can communicate. I haven't tried it with you version, but there are a few options left like contacting the auhor of the plugin or using A1111 with it. Just wanted to tell you. 😊

**Reply by FullBleed** [reply]
> > Sorry, I'm just a User who uses this plugin. 🙄 I absolutely understand it from your point of view, and I guess this is something coming from Adobe so that PS and WebUI can communicate. I haven't tried it with you version, but there are a few options left like contacting the auhor of the plugin or using A1111 with it. Just wanted to tell you. 😊

Please check with the author of the plugin to see how necessary it is.  We all should be much more careful about what (unnecessary) permissions we're giving out.

**Reply by vladmandic** [reply]
Maybe you can reach out and loop him here?
I'd be welcome to helping out if I can...

**Reply by Woisek** [reply]
Allright, I will try to let him know, as soon your version is up and running for some testing.
